drivers/sensor/bh1749/bh1749.c

```diff
@@ -0,0 +1,310 @@
+}
+
+static int bh1749_channel_get(struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct bh1749_data *data = dev->driver_data;
+
+	switch (chan) {
+	case SENSOR_CHAN_RED:
+		val->val1 = sys_le16_to_cpu(
+			data->sample_rgb_ir[BH1749_SAMPLE_POS_RED]);
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_GREEN:
+		val->val1 = sys_le16_to_cpu(
+			data->sample_rgb_ir[BH1749_SAMPLE_POS_GREEN]);
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_BLUE:
+		val->val1 = sys_le16_to_cpu(
+			data->sample_rgb_ir[BH1749_SAMPLE_POS_BLUE]);
+		val->val2 = 0;
+		break;
+	case SENSOR_CHAN_IR:
+		val->val1 = sys_le16_to_cpu(
+			data->sample_rgb_ir[BH1749_SAMPLE_POS_IR]);
+		val->val2 = 0;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int bh1749_check(struct bh1749_data *data)
+{
+	u8_t manufacturer_id;
+
+	if (i2c_reg_read_byte(data->i2c, DT_INST_0_ROHM_BH1749_BASE_ADDRESS,
+			      BH1749_MANUFACTURER_ID, &manufacturer_id)) {
+		LOG_ERR("Failed when reading manufacturer ID");
+		return -EIO;
+	}
+
+	LOG_DBG("Manufacturer ID: 0x%02x", manufacturer_id);
+
+	if (manufacturer_id != BH1749_MANUFACTURER_ID_DEFAULT) {
+		LOG_ERR("Invalid manufacturer ID: 0x%02x", manufacturer_id);
+		return -EIO;
+	}
+
+	u8_t part_id;
+
+	if (i2c_reg_read_byte(data->i2c, DT_INST_0_ROHM_BH1749_BASE_ADDRESS,
+			      BH1749_SYSTEM_CONTROL, &part_id)) {
+		LOG_ERR("Failed when reading part ID");
+		return -EIO;
+	}
+
+	if ((part_id & BH1749_SYSTEM_CONTROL_PART_ID_Msk) !=
+	    BH1749_SYSTEM_CONTROL_PART_ID) {
+		LOG_ERR("Invalid part ID: 0x%02x", part_id);
+		return -EIO;
+	}
+
+	LOG_DBG("Part ID: 0x%02x", part_id);
+
+	return 0;
+}
+
+static int bh1749_sw_reset(struct device *dev)
+{
+	return i2c_reg_update_byte(dev, DT_INST_0_ROHM_BH1749_BASE_ADDRESS,
+				   BH1749_SYSTEM_CONTROL,
+				   BH1749_SYSTEM_CONTROL_SW_RESET_Msk,
+				   BH1749_SYSTEM_CONTROL_SW_RESET);
+}
+
+static int bh1749_rgb_measurement_enable(struct bh1749_data *data, bool enable)
+{
+	u8_t en = enable ?
+		  BH1749_MODE_CONTROL2_RGB_EN_ENABLE :
+		  BH1749_MODE_CONTROL2_RGB_EN_DISABLE;
+
+	return i2c_reg_update_byte(data->i2c,
+				DT_INST_0_ROHM_BH1749_BASE_ADDRESS,
+			   BH1749_MODE_CONTROL2,
+			   BH1749_MODE_CONTROL2_RGB_EN_Msk,
+			   en);
+}
+
+
+static void bh1749_async_init(struct k_work *work)
+{
+	struct bh1749_data *data = bh1749_dev->driver_data;
+
+	ARG_UNUSED(work);
+
+
+
+	LOG_DBG("BH1749 async init step %d", data->async_init_step);
+
+	data->err = async_init_fn[data->async_init_step](data);
+
+	if (data->err) {
+		LOG_ERR("BH1749 initialization failed");
+	} else {
+		data->async_init_step++;
+
+		if (data->async_init_step == ASYNC_INIT_STEP_COUNT) {
+			LOG_INF("BH1749 initialized");
+		} else {
+			k_delayed_work_submit(&bh1749_init_work,
+				async_init_delay[data->async_init_step]);
+		}
+	}
+}
+
+static int bh1749_async_init_i2c(struct bh1749_data *data)
+{
+	data->i2c = device_get_binding(DT_INST_0_ROHM_BH1749_BUS_NAME);
```

Comment:
No need to do it in an async work step. It would be actually better to call it directly from `bh1749_init`, so that this function can return an error when the binding of the I2C device fails (as this a problem with configuration of the firmware, not in communication with the sensor or its operation).

Reply:
@anangl Moved to init